src/core/pal/labelposition.cpp

```diff
@@ -236,46 +234,99 @@ bool LabelPosition::isInConflict( const LabelPosition *lp ) const
   if ( this->probFeat == lp->probFeat ) // bugfix #1
     return false; // always overlapping itself !
 
+  const double minDuplicateSeparation = std::max( getFeaturePart()->feature()->thinningSettings().noRepeatDistance(),
+                                        lp->getFeaturePart()->feature()->thinningSettings().noRepeatDistance() );
+  const bool noDuplicateDistanceApplies = minDuplicateSeparation > 0 && getFeaturePart()->feature()->labelText() == lp->getFeaturePart()->feature()->labelText();
+
   // if either this label doesn't cause collisions, or the other one doesn't, then we don't conflict!
-  if ( this->feature->feature()->overlapHandling() == Qgis::LabelOverlapHandling::AllowOverlapAtNoCost ||
-       lp->feature->feature()->overlapHandling() == Qgis::LabelOverlapHandling::AllowOverlapAtNoCost )
+  const bool allowOverlapAtNoCost = this->feature->feature()->overlapHandling() == Qgis::LabelOverlapHandling::AllowOverlapAtNoCost ||
+                                    lp->feature->feature()->overlapHandling() == Qgis::LabelOverlapHandling::AllowOverlapAtNoCost;
+
+  // early exit shortcut
+  if ( !noDuplicateDistanceApplies && allowOverlapAtNoCost )
     return false;
 
-  if ( !nextPart() && !lp->nextPart() )
+  GEOSContextHandle_t geosctxt = QgsGeosContext::get();
+
+  const bool bothLabelsAreSinglePart = !nextPart() && !lp->nextPart();
```

Comment:
Nice use of const bool across the code to explain an otherwise hard to understand piece of QGIS :pray: